Nanosecond precision support, Integration tests for Timestream for LiveAnalytics migrations

fredjoonpark opened the pull request:
Issue #, if available:

N/A.

Description of changes:

- Integration tests have been added that run all scripts end-to-end.
- All scripts now use absolute import paths, meaning they can be run from any directory.
- All scripts have a new `main()` function that accepts a list of arguments. This enables testing and a single wrapper script in the future.
- `validator.py` now returns exit code `1` if any exceptions occur or if the counts don't match.
- The method `sync_line_protocol_to_storage` has been added to `s3_utils.py` and allows line protocol files to be downloaded to a local directory. This is crucial for integration tests to work.
- A `test_scripts` directory has been added, with scripts that set up a local InfluxDB Docker container for testing.
- `timestream-export-logs` added to `.gitignore`.
- Ingestion integration tests moved to `tests/integration/`, in the file `test_ingestion.py`.
- New location of `test-data` added to `.gitignore`.
- `time` column preserved in Athena table.
- Glue used to create Athena table, avoiding errors when one column name contains `,` and another column name contains `::`.
- Tests for nanosecond timestamp precision added.
- Tests for special characters included.
- Validation script updated to accept tag names with commas, provided they are surround by quotes.
- A new flag is added to the unload script (`--append-timestamps`). This flag appends an extra column for each measure of type timestamp to preserve nanosecond precision when migrating to InfluxDB, as <col_name>_ns.
- multi-measure record type `timestamp` is now supported for InfluxDB migrations

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

fredjoonpark committed:
adding ns timestream precision support

fredjoonpark committed:
docs

fredjoonpark committed:
cleanup

fredjoonpark committed:
docs

trevorbonas commented on tools/python/liveanalytics_migration_scripts/unload/utils/timestream_utils.py:
```diff
@@ -389,6 +389,8 @@ def build_query(self, migration_tag, database, table, bucket_s3_uri, partition,
         self.logger.info(f"Building unload query for {database}.{table}")
         unload_query = "UNLOAD("
         unload_query += " SELECT *"
+        if add_time_ns:
```
In my opinion, nanosecond precision should be used by default across all scripts.

fredjoonpark replied:
I agree, and we can explicitly recommend in the unload README that influxdb migrations should use this flag (since users expect nanosecond precision coming from LA). This also doesn't affect the original db/table so it should be safe, do you see any issues with setting this to be default @forestmvey @bobigbal?

forestmvey replied:
Agreed, let's make adding the nanoseconds column the default behavior.

fredjoonpark replied:
I have updated the main flow (unload with target: influxdb) to support nanosecond precision by default. Unload will now add a time_ns column on export, which will be stored in the parquet as a STRING. This column will then be used (instead of the time column) when translating to LP. If UNLOAD was performed without time_ns, there will be lossy timestamp translation (from Athena's limited ms support).

The change now uses pyarrow to just get the column details from an unloaded parquet (without performing reads). We could extend this logic to the LP translation if we like this approach.

forestmvey commented on tools/python/liveanalytics_migration_scripts/unload/unload.py:
```diff
@@ -45,6 +45,7 @@
     parser.add_argument("-cp", "--custom-partition-count", help="Custom partition count", default=99, required=False)
     parser.add_argument("-ob", "--order-by-asc", help="data order by time ascending", default=False, type=lambda x: x.lower() in ['true', '1', 'yes'], required=False)
     parser.add_argument("-ld", "--logs-dir", help='Directory for export logs (default: timestream-export-logs)', default = None, required = False)
+    parser.add_argument("-atn", "--add-time-ns", help="Add time_ns column in export to preserve timestamp precision for migrations to InfluxDB", default=False, type=lambda x: x.lower() in ['true', '1', 'yes'], required=False)
```
Did you notice a large difference in unload time when adding the additional column.

fredjoonpark replied:
for a dataset with 130 million rows, on average (over 3 runs):
- without time_ns: ~3m10s
- with time_ns: ~3m40s

a minor but slightly noticeable difference

forestmvey commented on tools/python/liveanalytics_migration_scripts/targets/timestream_for_influxdb/transform/transform.py:
```diff
@@ -643,6 +654,14 @@ def parse_table_dimensions(arg):
         required=True,
         type=parse_bool_cli_argument,
     )
+    parser.add_argument(
```
Rather than using a flag, why not dynamically determine if the `time_ns` column exists and make it the default. I don't see any reason why we would want to use the `time` column if the `time_ns` column already exists.

fredjoonpark committed:
making nanosecond precision be default

fredjoonpark committed:
cleanup

trevorbonas commented on tools/python/liveanalytics_migration_scripts/targets/timestream_for_influxdb/ingestion/README.md:
```diff
@@ -59,6 +59,7 @@ python influxdb_ingestion.py <bucket_name> <data_directory> [options]
 - `-l, --lines`: Number of lines per batch (default: 10000)
 - `-m, --multiplier`: I/O multiplier - how many batches to read at once (default: 10)
 - `-r, --retries`: Maximum retry attempts for failed batches (default: 20)
+- `-p, --precision`: Timestamp precision for InfluxDB write (default: ns)
```
We should describe what [options are available for precision](https://docs.influxdata.com/influxdb/v2/api/v2/#tag/Write) since it's not straightforward. For example, `us` is microseconds.

trevorbonas commented on tools/python/liveanalytics_migration_scripts/targets/timestream_for_influxdb/ingestion/influxdb_ingestion.py:
```diff
@@ -458,6 +459,9 @@ def main():
     parser.add_argument('--continue-on-error', action='store_true',
                         help='Continue ingesting remaining files even if one fails')
+    parser.add_argument('-p', '--precision', type=str, default='ns',
+                        choices=['ns', 'ms'],
```
```suggestion
                        choices=['ns', 'ms', 'us', 's'],
```

trevorbonas commented on tools/python/liveanalytics_migration_scripts/unload/utils/s3_utils.py:
```diff
@@ -188,6 +188,46 @@ def get_latest_unload_path(
+                "bucket_path must be an S3 URI of the form s3://<bucket>/<prefix>"
+            )
+
+        bucket_and_prefix = bucket_path[5:]
```
Why is the `s3://` scheme required for the bucket path when it is discarded?

trevorbonas commented on tools/python/liveanalytics_migration_scripts/unload/utils/s3_utils.py:
```diff
@@ -188,6 +188,46 @@ def get_latest_unload_path(
+
+        self.wait_for_multipart_uploads(bucket_name=bucket_name, prefix=prefix)
+
+        # Paginate through keys until we encounter the first *.file_type
```
Are these ordered in any particular way? Do we care about the order?

fredjoonpark replied:
We don't care about the order (any file will contain the schema)

trevorbonas commented on tools/python/liveanalytics_migration_scripts/targets/timestream_for_influxdb/ingestion/influxdb_ingestion.py:
```diff
@@ -233,7 +233,7 @@ def ingest_batch(write_api, batch, batch_id, process_name, bucket_name, max_retr
                 logging.warning(f"{process_name} - Retry attempt {ingestion_attempt} for batch {batch_id}")
 
-            write_api.write(bucket=bucket_name, record=batch_content, write_precision="ms")
+            write_api.write(bucket=bucket_name, record=batch_content, write_precision=precision)
```
Worth noting somewhere that if the precision here does not match the actual precision of the timestamp, for example, if data is unloaded with millisecond precision but ingested with `write_precision="ns"`, timestamps will be very off.

trevorbonas commented on tools/python/liveanalytics_migration_scripts/requirements.txt:
```diff
@@ -6,3 +6,5 @@ influxdb-client==1.48.0
 requests==2.32.3
 psycopg2-binary==2.9.10
+pyarrow==20.0.0
+s3fs==2025.5.0
```
When installing packages with this requirements.txt file, I get this error:
```
ERROR: Cannot install -r requirements.txt (line 1) and s3fs because these package versions have conflicting dependencies.

The conflict is caused by:
    boto3 1.38.0 depends on botocore<1.39.0 and >=1.38.0
    aiobotocore 2.22.0 depends on botocore<1.37.4 and >=1.37.2
    . . .
    boto3 1.38.0 depends on botocore<1.39.0 and >=1.38.0
    aiobotocore 2.5.4 depends on botocore<1.31.18 and >=1.31.17
```

trevorbonas replied:
The only thing that fixed this for me was removing `boto3` and adding `boto3` as an `s3fs` extra.
```
s3fs[boto3]==2025.5.0
```

trevorbonas commented on tools/python/liveanalytics_migration_scripts/targets/timestream_for_influxdb/transform/transform.py:
```diff
@@ -127,9 +120,22 @@ def get_athena_ddl_type_mapping(timestream_type: str) -> str:
+            formatted_columns.append(f"`{field.name}` {column_type}")
+        return formatted_columns
+    except Exception as e:
+        raise
```
If nothing is done with `e` and the exception is passed along, this `try` `except` block isn't necessary.

fredjoonpark committed:
using pyarrow fs

trevorbonas commented on tools/python/liveanalytics_migration_scripts/unload/utils/s3_utils.py:
```diff
@@ -188,6 +188,46 @@ def get_latest_unload_path(
+                `s3://my-bucket/benchmark22/cpu/unload-2025-05-23-18:58:58/results`
+
+        Returns:
+            str: Full S3 URI to the first file type (ie. parquet) object discovered.
```
```suggestion
            str: Full S3 URI to the first file type (i.e., parquet) object discovered.
```

trevorbonas commented on tools/python/liveanalytics_migration_scripts/unload/utils/s3_utils.py:
```diff
@@ -188,6 +188,46 @@ def get_latest_unload_path(
+        supplied bucket_path prefix.
+
+        Args:
+            bucket_path (str): An S3 URI including bucket name and prefix ie.
```
```suggestion
            bucket_path (str): An S3 URI including bucket name and prefix, i.e.,
```

fredjoonpark replied:
Addressed [in #57](https://github.com/Bit-Quill/amazon-timestream-tools/pull/57)

forestmvey commented on tools/python/liveanalytics_migration_scripts/targets/timestream_for_influxdb/transform/transform.py:
```diff
@@ -689,34 +668,44 @@ def parse_table_dimensions(arg):
-            add_validation_field=args.add_validation_field,
-        )
-        line_protocol_results.append(line_protocol_result)
+            # TODO: Update to retrieve schema from parquet
```
Is there any reason we don't want to scope this into the current PR?

fredjoonpark replied:
upon further review, the transformation will need to query Timestream to distinguish tags and dimensions. Removed the TODO 👍

forestmvey reviewed (approved):
Tested and validated nanoseconds timestamps are preserved through `unload` &rarr; `transformation` 👍

fredjoonpark committed:
Support preserving nanosecond precision for all timestamp multi-measure record types (#57)

* support precision for all ts columns
* adding bigint support, ensure integer conversions

trevorbonas commented on tools/python/liveanalytics_migration_scripts/targets/timestream_for_influxdb/transform/transform.py:
```diff
@@ -127,9 +120,44 @@ def get_athena_ddl_type_mapping(timestream_type: str) -> str:
+                ns_field_name = f"{field_name}_ns"
+                # Check if corresponding _ns field exists
+                if ns_field_name in all_field_names:
+                    # Use the _ns field instead and mark both as processed
```
This will cause `time` to be absent in Athena, causing validation to fail.

trevorbonas commented on tools/python/liveanalytics_migration_scripts/targets/timestream_for_influxdb/transform/transform.py:
```diff
@@ -127,9 +120,44 @@ def get_athena_ddl_type_mapping(timestream_type: str) -> str:
+                    # No _ns field exists, use the original field
+                    formatted_columns.append(f"`{field_name}` TIMESTAMP")
+                    processed_fields.add(field_name)
+            elif column_type == "INT64":
```
The `BOOL` Arrow data type, among others, is not handled.

trevorbonas commented:
I have added fixes for Parquet file schema retrieval in the integration test branch: https://github.com/Bit-Quill/amazon-timestream-tools/pull/55/files#diff-1e6cd5d2aa4f0d314adbf5583322216492e3fc7d81e2c1eb08d16c9ccc904239

trevorbonas committed:
Add migration scripts integration tests (#55)

Changes:
- Integration tests have been added that run all scripts end-to-end.
- All scripts now use absolute import paths, meaning they can be run from any directory.
- All scripts have a new `main()` function that accepts a list of arguments. This enables testing and a single wrapper script in the future.
- `validator.py` now returns exit code `1` if any exceptions occur or if the counts don't match.
- The method `sync_line_protocol_to_storage` has been added to `s3_utils.py` and allows line protocol files to be downloaded to a local directory. This is crucial for integration tests to work.
- A `test_scripts` directory has been added, with scripts that set up a local InfluxDB Docker container for testing.
- `timestream-export-logs` added to `.gitignore`.
- Ingestion integration tests moved to `tests/integration/`, in the file `test_ingestion.py`.
- New location of `test-data` added to `.gitignore`.
- `time` column preserved in Athena table.
- Glue used to create Athena table, avoiding errors when one column name contains `,` and another column name contains `::`.
- Tests for nanosecond timestamp precision added.
- Tests for special characters included.
- Validation script updated to accept tag names with commas, provided they are surround by quotes.

fredjoonpark committed:
location constraint only for non us east 1

fredjoonpark committed:
setting up athena database, increasing test timeouts
